### ISSUE TITLE
Feature request: Accept iterables and async iterables

### ISSUE THREAD

**Issue by m93a** [supercharge/promise-pool#71]
Currently, the PromisePool.for method only accepts an array of items. However, there are many cases where it would be helpful to also support iterables and async iterables as input.

For example, my use case involves scraping a paginated web page using [puppeteer](https://pptr.dev/), where I asynchronously yield the article urls. I would like to use a promise pool to fetch the articles, but since I am using an async iterable to generate the URLs, I am unable to use PromisePool.for in its current form.

I believe that allowing PromisePool to accept iterables and async iterables would greatly increase the flexibility and usefulness of this library, and I would be happy to contribute a pull request to add this feature.

There is some overlap with the issue #65, which proposed allowing new items to be added to the pool. I believe that once the logic for consuming iterables is implemented, it won't be hard to also add methods like `queueNext` and `queueLast` that would add new items to the pool's queue.

**Comment by marcuspoehls** [comment]
@m93a Hi Michal, that's a nice idea. I'd love to add support for iterables and appreciate your support by submitting a pull request 👍 please go ahead and create a PR. Thank you!

**Comment by m93a** [comment]
Hi! I'm having problems with running the tests. I cloned the repo, ran `pnpm install` and then `pnpm test`, and the tests failed on the following TypeScript errors:
```
src/promise-pool-error.ts:28:11 - error TS2339: Property 'captureStackTrace' does not exist on type 'ErrorConstructor'.

28     Error.captureStackTrace(this, this.constructor)
             ~~~~~~~~~~~~~~~~~

src/promise-pool-executor.ts:536:7 - error TS2304: Cannot find name 'setTimeout'.

536       setTimeout(() => {
          ~~~~~~~~~~

src/validation-error.ts:12:11 - error TS2339: Property 'captureStackTrace' does not exist on type 'ErrorConstructor'.

12     Error.captureStackTrace(this, this.constructor)
             ~~~~~~~~~~~~~~~~~


Found 3 errors in 3 files.
```
The same happens if I run `tsc` in the project's root directory. Is there a lib missing or something?

**Comment by marcuspoehls** [comment]
Hey Michal, that’s strange. The test suite works fine on my machine and here on GitHub Actions without adding the extra types. What version of Node.js are you using? I guess something like 16.x or above. If adding `@types/node` fixes it for you, we’ll add it as a dev dependency :) there are likely other people that would run into this issue

**Comment by m93a** [comment]
I tested with Node v19 as well as v14, and with NPM as well as PNPM. Turns out that the problem was caused by me using PNPM at first! If I install the dependencies using NPM, everything works fine even in Node v19. I guess NPM and PNPM resolve the package versions differently or something.

**Comment by marcuspoehls** [comment]
Good to know, thank you for sharing the details 👍

**Comment by marcuspoehls** [comment]
@m93a Hey Michal, I’ve just published version `3.0.0` to NPM 😃 thank you very much for your work on supporting iterables and async iterables 👍
